src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java

```diff
@@ -189,4 +190,109 @@ private boolean isResponseOutputEqual(SessionResultsData.ResponseOutput self,
+        loginAsStudent(student1InCourse1.googleId);
+        verifyCanAccess(submissionParams);
+    }
+
+    @Test
+    public void testAccessControl_studentAccessUnpublishedSessionStudentResult_shouldFail() {
+        StudentAttributes student1InCourse1 = typicalBundle.students.get("student1InCourse1");
+        CourseAttributes typicalCourse1 = typicalBundle.courses.get("typicalCourse1");
+        FeedbackSessionAttributes feedbackSessionAttributes = typicalBundle.feedbackSessions.get("session2InCourse1");
+
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.COURSE_ID, typicalCourse1.getId(),
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionAttributes.getFeedbackSessionName(),
+                Const.ParamsNames.INTENT, Intent.STUDENT_RESULT.toString(),
+        };
+        loginAsStudent(student1InCourse1.googleId);
+        verifyCannotAccess(submissionParams);
+    }
+
+    @Test
+    public void testAccessControl_accessStudentSessionResultWithMasqueradeMode_shouldPass() {
+        StudentAttributes student1InCourse1 = typicalBundle.students.get("student1InCourse1");
+        CourseAttributes typicalCourse1 = typicalBundle.courses.get("typicalCourse1");
+        FeedbackSessionAttributes feedbackSessionAttributes = typicalBundle.feedbackSessions.get("session1InCourse1");
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.COURSE_ID, typicalCourse1.getId(),
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionAttributes.getFeedbackSessionName(),
+                Const.ParamsNames.INTENT, Intent.STUDENT_RESULT.toString(),
+        };
+
+        publishSession(feedbackSessionAttributes.getFeedbackSessionName(), typicalCourse1.getId());
+        loginAsAdmin();
+        verifyCanMasquerade(student1InCourse1.googleId, submissionParams);
+    }
+
+    @Test
+    public void testAccessControl_studentAccessOtherCourseSessionResult_shouldFail() {
+        StudentAttributes student1InCourse1 = typicalBundle.students.get("student1InCourse1");
+        CourseAttributes typicalCourse1 = typicalBundle.courses.get("typicalCourse1");
+        CourseAttributes typicalCourse2 = typicalBundle.courses.get("typicalCourse2");
+        FeedbackSessionAttributes feedbackSessionAttributes = typicalBundle.feedbackSessions.get("session1InCourse2");
+
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.COURSE_ID, typicalCourse2.getId(),
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionAttributes.getFeedbackSessionName(),
+                Const.ParamsNames.INTENT, Intent.STUDENT_RESULT.toString(),
+        };
+
+        loginAsStudent(student1InCourse1.googleId);
+        verifyCannotAccess(submissionParams);
+
+        // Malicious api call using course Id of the student to bypass the check
+        submissionParams[1] = typicalCourse1.getId();
+        verifyEntityNotFound(submissionParams);
```

Comment:
```suggestion
```

The CI is failing because this test is not passing. Actually, I don't think this should be an issue any longer ever since the logic for `GetSessionResultsAction` was refactored. @xpdavid can you help me confirm this?

Reply:
Nope. The refactor does not touch the `checkSpecificAccessControl`. The root cause of the failing test is that we are testing the wrong method: We should test `checkSpecificAccessControl` not `execute`. `verifyEntityNotFound ` wrongly call the `execute` method.